packages/codec/lib/format/utils/exception.ts

```diff
@@ -0,0 +1,147 @@
+import debugModule from "debug";
+const debug = debugModule("codec:utils:errors");
+
+import * as Format from "@truffle/codec/format/common";
+import * as Ast from "@truffle/codec/ast";
+import * as Storage from "@truffle/codec/storage/types";
+
+function isContractDefinedType(
+  anyType: Format.Types.Type
+): anyType is Format.Types.ContractDefinedType {
+  const contractDefinedTypes = ["enum", "struct"];
+  return contractDefinedTypes.includes(anyType.typeClass);
+}
+
+//this function gives an error message
+//for those errors that are meant to possibly
+//be wrapped in a DecodingError and thrown
+export function message(error: Format.Errors.ErrorForThrowing): string {
+  switch (error.kind) {
+    case "UserDefinedTypeNotFoundError":
+      let typeName = isContractDefinedType(error.type)
+        ? error.type.definingContractName + "." + error.type.typeName
+        : error.type.typeName;
+      return `Unknown ${error.type.typeClass} type ${typeName} of id ${
+        error.type.id
+      }`;
+    case "UnsupportedConstantError":
+      return `Unsupported constant type ${Ast.Utils.typeClass(
+        error.definition
+      )}`;
+    case "ReadErrorStack":
+      return `Can't read stack from position ${error.from} to ${error.to}`;
+    case "ReadErrorBytes":
+      return `Can't read ${error.length} bytes from input starting at ${
+        error.start
+      }`;
+    case "ReadErrorStorage":
+      if (error.range.length) {
+        return `Can't read ${
+          error.range.length
+        } bytes from storage starting at index ${
+          error.range.from.index
+        } in ${slotAddressPrintout(error.range.from.slot)}`;
+      } else {
+        return `Can't read storage from index ${
+          error.range.from.index
+        } in ${slotAddressPrintout(error.range.from.slot)} to index ${
+          error.range.to.index
+        } in ${slotAddressPrintout(error.range.to.slot)}`;
+      }
+  }
+}
+
+function slotAddressPrintout(slot: Storage.Slot): string {
```

Comment:
Does this function really belong here either?  I mean, I can see the logic, this is the only place it gets used, so I guess it makes sense, but it still seems a little out of place.

Reply:
(Same reasoning as `isContractDefinedType`)

Anyway, for these decisions I made to un-export things, I did so mostly out of practicality, figuring we can move and/or export it later when we need it somewhere else.

Reply:
Yeah I mean this is a fine place for it for now.